src/Xpress.jl

```diff
@@ -1,99 +1,119 @@
-
-    include("xprs_vars.jl")
-
-    include("xprs_constrs.jl")
-    include("xprs_quad.jl")
-    include("xprs_highlevel.jl")
-
-    include("xprs_solve.jl")
-    include("xprs_callbacks.jl")
-    include("xprs_iis.jl")
-
-    include("XpressSolverInterface.jl")
-    include("MOIWrapper.jl")
-
-    # license checker
-    include("xprs_userlic.jl")
-
-    function __init__()
-
-        # some lics require special check
-        userlic()
-
-        # start Xpress with XPRSinit
-        Env()
-    end
-end
+__precompile__()
+
+module Xpress
+
+    using Libdl
+
+    # Load in `deps.jl`, complaining if it does not exist
+    const depsjl_path = joinpath(@__DIR__, "..", "deps", "deps.jl")
+
+    if !isfile(depsjl_path)
+        error("XPRESS cannot be loaded. Please run Pkg.build(\"Xpress\").")
+    end
+
+    include(depsjl_path)
+
+    ### imports
+
+    import Base.show, Base.copy
+
+    # ### exports
+    # export
+
+    # ## xprs_env
+    # free_env,
+
+    # ## xprs_params
+    # getparam, setparam!, setparams!,
+
+    # ## xprs_model
+    # set_sense!, update_model!, reset_model!,
+    # read_model, write_model,
+    # copy,
+
+    # ## xprs_attrs
+    # model_name, model_sense, model_type,
+    # num_vars, num_constrs, num_sos, num_qconstrs,
+    # num_cnzs, num_qnzs, num_qcnzs,
+    # is_qp, is_qcp, is_mip,
+
+    # lowerbounds, upperbounds, objcoeffs, set_objcoeffs!,
+
+    # ## xprs_vars
+    # XPRS_CONTINUOUS, XPRS_BINARY, XPRS_INTEGER,
+    # add_var!, add_vars!, add_cvar!, add_cvars!,
+    # add_bvar!, add_bvars!, add_ivar!, add_ivars!,
+    # del_vars!,
+
+    # ## xprs_constrs
+    # add_constr!, add_constrs!, add_constrs_t!,
+    # add_rangeconstr!, add_rangeconstrs!, add_rangeconstrs_t!,
+    # get_constrmatrix, add_sos!, del_constrs!, chg_coeffs!,
+
+    # ## xprs_quad
+    # add_qpterms!, add_qconstr!,
+
+    # ## higher level
+    # xpress_model,
+
+    # ## xprs_solve
+    # optimize, computeIIS, get_solution,
+    # get_status, OptimInfo, get_optiminfo, get_objval
+
+
+    ### include source files
+     module Lib
+
+        using CEnum
+
+        import ..Xpress
+        const libxprs = Xpress.libxprs
+
+        include("ctypes.jl")
+        include("common.jl")
+        include("lib.jl")
+
+        # include("custom.jl")
+
+    end
+
+    include("utils.jl")
+
+    include("helper.jl")
+
+    include("api.jl")
+
+    # # license checker
+    include("license.jl")
+
+    # include("xprs_full_defines.jl")
+
+    # include("xprs_common.jl")
+    # include("xprs_env.jl")
+
+    # include("xprs_model.jl")
+    # include("xprs_params.jl")
+    # include("xprs_attrs.jl")
+
+    # include("xprs_vars.jl")
+
+    # include("xprs_constrs.jl")
+    # include("xprs_quad.jl")
+    # include("xprs_highlevel.jl")
+
+    # include("xprs_solve.jl")
+    # include("xprs_callbacks.jl")
+    # include("xprs_iis.jl")
+
+    # include("XpressSolverInterface.jl")
+    # include("MOIWrapper.jl")
+
+    function __init__()
```

Comment:
This is not related to the re write.
But I was thinking about removing this and forcing the user to call the functions explicitly.
One intermediary option would be to keep a global Vector variable “is_initialized” That is set by init. In this case the mains proble constructor could call it if false.

Reply:
You mean `XPRSinit`, right? I prefer that too. In fact in my local copy I don't have that and the license check in the `__init__`, but I just don't commit that patch.

Tradeoffs as I see it: 

- We need to have XPRSinit called and run successfully before anyone can use this package. Currently someone can run `using Xpress`, have it fail and type `using Xpress` again, and since Julia already ran `__init__` it won't run it again. This might give the impression to a user that the package initialized successfully when they import the second time. I've been in this situation. If anything, we can call it in `Xpress` module scope. If it errors, typing `using Xpress` again will force precompilation.
- If we don't call XPRSinit during import or initialization any time a user calls any Xpress library function we should throw the appropriate error. So we have to run a check before every function call (I don't like this), or run a check when an error is thrown (I prefer this).

Reply:
Lets go for: not calling on initialization.
This "run a check when an error is thrown (I prefer this)." is reasonable.
But I think almost all functions depend on a `XPRSprob`, we might consider testing the initialization inside that constructor.
So maybe check in constructor and when error is thrown?

Reply:
Resolved in https://github.com/JuliaOpt/Xpress.jl/pull/43/commits/8498ec359d03f369b31c2bf6fcd0143af82d59ab.

```
$ julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.3.0-rc4.1 (2019-10-15)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]

julia> Xpress.getversion()
v"29.1.10"

julia> Xpress.getbanner()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
ERROR: XpressError: Unrecoverable error. Unable to invoke XPRSgetbanner
Stacktrace:
 [1] invoke(::Function, ::Int64, ::Type{String}) at /Users/USER/GitRepos/Xpress.jl/src/utils.jl:22
 [2] getbanner() at /Users/USER/GitRepos/Xpress.jl/src/api.jl:257
 [3] top-level scope at REPL[3]:1

julia> xp = Xpress.XpressProblem()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
ERROR: Unable to create a Xpress Problem. Received error code 255
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] Xpress.XpressProblem() at /Users/USER/GitRepos/Xpress.jl/src/helper.jl:56
 [3] top-level scope at REPL[4]:1

julia> Xpress.initialize()
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> Xpress.getbanner()
"FICO Xpress-Optimizer 64-bit v29.01.10 (Hyper capacity)\n(c) Copyright Fair Isaac Corporation 1983-2016. All rights reserved"

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x00000001125f5000)

julia>


```